src/dmqproto/node/neo/request/Consume.d

```diff
@@ -1004,3 +1004,469 @@ public abstract scope class ConsumeProtocol_v2
+        }
+    }
+
+    /***************************************************************************
+
+        Called when the storage is empty. Waits for
+          - the `Push` signal,
+          - the `Flush` signal, if an incomplete batch is pending,
+          - a `Stop` message from the client and acknowledges it by sending a
+            `Stopped` message.
+
+        Returns:
+            - `NodeFiberResumeCode.Pushed` or `NodeFiberResumeCode.Flush` if
+              resumed with one of these signals,
+            - 0 if a `Stop` message from the client arrived.
+
+    ***************************************************************************/
+
+    private uint wait ( )
+    {
+        this.resume_fiber_on_push = true;
+        this.resume_fiber_on_flush = !!(*this.record_batch).length;
+
+        auto event = this.ed.nextEvent(ed.NextEventFlags.Receive);
+
+        this.resume_fiber_on_push = false;
+        this.resume_fiber_on_flush = false;
+
+        switch (event.active)
+        {
+            case event.active.resumed:
+                return event.resumed.code;
+
+            case event.active.received:
+                // Received message while the DMQ channel is empty: It should
+                // be Stop. Acknowledge Stop and return 0.
+                this.verifyReceivedMessageIsStop(event.received.payload);
+                this.sendStoppedMessage();
+                static assert(NodeFiberResumeCode.min > 0);
+                return 0;
+
+            default:
+                assert(false, "Consume: Unexpected fiber resume code");
+        }
+    }
+
+    /***************************************************************************
+
+        Sends the current batch of records; that is, `*this.record_batch`, and
+        waits for a `Continue` or `Stop` message. Clears `*this.record_batch`
+        when finished.
+
+    ***************************************************************************/
+
+    private bool sendBatchAndReceiveFeedback ( )
+    {
+        void fillinRecordsMessage ( ed.Payload payload )
+        {
+            payload.addConstant(MessageType_v3.Records);
+            payload.addArray(*this.record_batch);
+        }
+
+        scope (exit)
+        {
+            (*this.record_batch).length = 0;
+            enableStomping(*this.record_batch);
+        }
+
+        // Send the records but be ready to potentially receive a Stop message.
+        auto event = this.ed.nextEvent(
+            ed.NextEventFlags.Receive, &fillinRecordsMessage
+        );
+
+        switch (event.active)
+        {
+            case event.active.sent:
+                // Records sent: Wait for Consume/Stop feedback, acknowledge
+                // Stop and return true for Continue or false for Stop.
+                switch (this.ed.receiveValue!(MessageType_v3)())
+                {
+                    case MessageType_v3.Continue:
+                        return true;
+
+                    case MessageType_v3.Stop:
```

Comment:
To be honest, I think you'd find the logic of this class would be simplified by using the `RequestEventDispatcher`. The fact that you have to handle receiving the Stop message in two locations shows that there are two processes going on in parallel:
1. Popping from the channel, building up batches, sending them to the client, then waiting for the Continue message.
2. Handling Stop messages from the client.